man/refile.man
.TH REFILE %manext1% 2013-03-19 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
refile \- file message in nmh folders
.SH SYNOPSIS
.HP 5
.na
.B refile
.RB [ \-help ]
.RB [ \-version ]
.RI [ msgs ]
.RB [ \-draft ]
.RB [ \-link " | " \-nolink ]
.RB [ \-preserve " | " \-nopreserve ]
.RB [ \-retainsequences " | " \-noretainsequences ]
.RB [ \-unlink " | " \-nounlink ]
.RB [ \-src
.IR +folder ]
.RB [ \-file
.IR file ]
.RB [ \-rmmproc
.IR program ]
.RB [ \-normmproc ]
.I +folder1
\&...
.ad
.SH DESCRIPTION
.B refile
moves (see
.IR mv (1))
or links (see
.IR ln (1))
messages from a source folder into one or more destination folders.
.PP
If you think of a message as a sheet of paper, this operation is not
unlike filing the sheet of paper (or copies) in file cabinet folders.
When a message is filed, it is linked into the destination folder(s)
if possible, and is copied otherwise.  As long as the destination
folders are all on the same file system, multiple filing causes little
storage overhead.  This facility provides a good way to cross-file or
multiple-index messages.  For example, if a message is received from
Jones about the ARPA Map Project, the command
.PP
.RS 5
refile\0cur\0+jones\0+Map
.RE
.PP
would allow the message to be found in either of the two folders `jones'
or `Map'.
.PP
You may specify the source folder using
.B \-src
.IR +folder .
If this is not given, the current folder is used by default.
If no message is specified, then `cur' is used by default.
.PP
The option
.B \-file
.I file
directs
.B refile
to use the specified file as the source message to be filed, rather
than a message from a folder.
Note that the file should be a validly formatted message, just like
any other
.B nmh
message.  It should
.I not
be in mail drop format (to convert a file in
mail drop format to a folder of
.B nmh
messages,
see
.IR inc (1)).
.PP
If a destination folder doesn't exist,
.B refile
will ask if you want to create it.
A negative response will abort the file operation.
If the standard input for
.B refile
is
.I not
a tty, then
.B refile
will not ask any questions and will proceed as if the user answered
\*(lqyes\*(rq to all questions.
.PP
The option
.B \-link
preserves the source folder copy of the message (i.e., it does an
.IR ln (1)
rather than a
.IR mv (1)),
whereas,
.B \-nolink
(the default) deletes the filed messages from the source folder.
.PP
Normally when a message is refiled, for each destination folder it
is assigned the number which is one above the current highest message
number in that folder.  Use of the
.B \-preserve
switch will override
this message renaming, and try to preserve the number of the message.
If a conflict for a particular folder occurs when using the
.B \-preserve
switch, then
.B refile
will use the next available message number which is above the message
number you wish to preserve.
.PP
As message sequences are folder-specific, moving the message from
the source folder removes it from all its sequences in that folder.
.B \-retainsequences
adds it to those same sequences in the destination folder,
creating any that don't exist.
This adding does not apply for the \*(lqcur\*(rq sequence.
.PP
If
.B \-link
is not specified (or
.B \-nolink
is specified), the filed messages will be removed from the source
folder.  The default is to remove these messages by renaming them
with a site-dependent prefix (usually a comma).
Such files will then need to be removed in some manner after a
certain amount of time.  Many sites arrange for
.B cron
to remove these files once a day, so check with your system administrator.
.PP
Alternately, if you wish for
.B refile
to really remove the files representing these messages from the source
folder, you can use the
.B \-unlink
switch (not to be confused with the
.B \-link
switch).  But messages removed by this method cannot be later recovered.
.PP
If you prefer a more sophisticated method of `removing' the messages
from the source folder, you can define the
.B rmmproc
profile component.  For example, you can add a profile component such as
.PP
.RS 5
rmmproc:    /home/coleman/bin/rmm_msgs
.RE
.PP
then
.B refile
will instead call the named program or script to
handle the message files.
.PP
The user may specify
.B \-rmmproc
.I program
on the command line to override this profile specification.
The
.B \-normmproc
option forces the message files to be deleted by renaming or
unlinking them as described above.
.PP
The
.B \-draft
switch tells
.B refile
to file the <mh-dir>/draft.
.SH FILES
.fc ^ ~
.nf
.ta \w'%nmhetcdir%/ExtraBigFileName  'u
^$HOME/.mh_profile~^The user profile
.fi
.SH "PROFILE COMPONENTS"
.fc ^ ~
.nf
.ta 2.4i
.ta \w'ExtraBigProfileName  'u
^Path:~^To determine the user's nmh directory
^Current\-Folder:~^To find the default current folder
^Folder\-Protect:~^To set mode when creating a new folder
^rmmproc:~^Program to delete the message
.fi
.SH "SEE ALSO"
.IR folder (1),
.IR mh-sequence (5),
.IR rmf (1),
.IR rmm (1)
.SH DEFAULTS
.nf
.RB ` "\-src\ +folder" "' defaults to the current folder"
.RB ` msgs "' defaults to cur"
.RB ` \-nolink '
.RB ` \-nounlink '
.RB ` \-nopreserve '
.RB ` \-noretainsequences '
.fi
.SH CONTEXT
If
.B \-src
.I +folder
is given, it will become the current folder.
If neither
.B \-link
nor `all' is specified, the current message in the
source folder will be set to the last message specified; otherwise,
the current message won't be changed.
.PP
If the \*(lqPrevious\-Sequence\*(rq profile entry is set, in addition
to defining the named sequences from the source folder,
.B refile
will also define those sequences for the destination folders.
See
.IR mh\-sequence (5)
for information concerning the previous sequence.
.SH BUGS
Since
.B refile
and
.B rmm
use your
.I rmmproc
to delete the message, the
.I rmmproc
must
.I not
call
.B refile
or
.B rmm
without specifying
.BR \-normmproc ,
or you will create an infinite loop.


NEWS
#
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

---------------
NOTABLE CHANGES
---------------

The largest notable changes in the 1.7 release are:

- Complete unification of network security support.  All network protocols
  (currently, POP and SMTP) have been refactored to use a common set of
  security routines.  This means all protocols support all SASL mechanisms
  (via the Cyrus-SASL library) and TLS.  TLS support has been strengthened
  to perform certificate name validation and to require TLS 1.1 as a
  minimum protocol.  Also, all protocols can make use of the OAuth2/XOAUTH
  SASL mechanism, which is supported by gmail.
- send(1) now supports adding switches to post(8) based on the address or
  domain of the email address in the From: header; this more easily allows
  users to support multiple identities.
- A generic facility for passing arguments to filter programs in repl(1)
  by use of the -convertargs switch.
- Native support for the manipulation of iCalendar requests; see mhical(1)
  for more details.

------------
NEW FEATURES
------------

The following are new features for the 1.7 release of nmh:

- When building from source, configure will derive ${prefix} from an existing
  nmh installation if it finds one in your $PATH.
- Added welcome message when nmh detects that its version changed.
- The default locations for configuration files and support binaries
  have been changed.  Configuration files now install into ${sysconfdir}/nmh,
  and support binaries are placed in ${libexecdir}/nmh.  If you are upgrading
  an existing installation you should look for old configuration files in
  ${sysconfdir} and merge any local customizations into the new files in
  ${sysconfdir}/nmh, then remove the old files.  ${libdir} will also contain
  obsolete support programs that should be removed.
- All TLS connections now perform certificate validation (including hostname
  matching) by default; can be disabled on a per-application basis.
- post now defaults to port 587 on 'smtp' message submission.
- A value of 0 for the width switch of scan(1), inc(1), ap(1), dp(1),
  fmttest(1), and mhl(1) now means as many characters as the format
  engine can produce [Bug #15274].  That amount is limited by internal
  buffers.
- If a component has trailing whitespace, e.g., body:component="> ",
  mhl now trims that whitespace off when filtering blank text lines.
- An "rtrim" flag has been added to mhl to remove any trailing
  whitespace from filtered text lines.  A corresponding "nortrim" flag
  has also been added.
- Added getmymbox and getmyaddr mh-format(5) function escapes.
- New -[no]changecur, -fixtype, -decodetypes, and -[no]crlflinebreaks switches
  have been added to mhfixmsg(1).
- mhfixmsg now removes an extraneous trailing semicolon from header
  parameter lists.
- Added -convertargs switch to repl(1), to pass arguments to programs
  specified in the user's profile or mhn.defaults to convert message
  content.
- Added mhical(1), to display, reply to, and cancel iCalendar (RFC 5545)
  event requests.
- Added multiply mh-format(5) function.
- "mhparam bindir" prints the path to the directory containing the public
  executables (${bindir}).
- New "-prefer" switch for mhshow (and mhlist and mhshow), to allow specifying
  the preferred content types to show, if present in a multipart alternative.
- mh-format now has %(kilo) and %(kibi) functions, to allow printing
  numbers with SI or IEC quantities, e.g. "10K", "2.3Mi".
- Support for the -sendmail flag to send/post to change the sendmail
  binary when using the sendmail/pipe MTS.
- Added support to send(1) to specify switches to post(1) based on address or
  domain name in From: header line in message draft.
- post(8) -snoop now attempts to decode base64-encoded SMTP traffic.
- folder(1) -nocreate now prints a warning message for a non-existent folder.
- mhfixmsg(1) now allows -decodetext binary, though 8bit is still the default.
- inc(1) and msgchk(1) now support TLS encryption natively.
- All network protocols support the XOAUTH authentication mechanism.
- Support for SMTPUTF8 (RFC 6531) has been added.  mhshow(1) already supported
  RFC 6532, assuming all 8-bit message header field bodies are UTF-8 and use
  of a UTF-8 locale.
- mhfixmsg now replaces RFC 2047 encoding with RFC 2231 encoding of name and
  filename parameters in Content-Type and Content-Disposition headers,
  respectively.
- If a message body contains 8-bit bytes, post(8) uses SMTP 8BITMIME if the
  server supports it.  If not, post fails with a message to the user to
  encode the message for 7-bit transport.
- Fewer lseek(2)s will be used when reading headers in the common case.
- ./configure's --enable-debug has been removed;  it did nothing.
- configure now defaults to enabling each of TLS and Cyrus SASL if the
  necessary headers and libraries are found.
- Moved build_nmh to top-level directory.
- Better error reporting for connections to network services.

---------
BUG FIXES
---------

- The format scanner no longer subtracts 1 from the width.  This has the
  effect of no longer counting the trailing newline in the output of
  scan(1), inc(1), and the other programs that rely on the format scanner.
- The first character of some very short (less than 4 characters) message
  bodies is no longer dropped.
- Single-character headers can be reliably formatted, etc., instead of
  apparently being missing.
- mhfixmsg now adds a Content-Transfer-Encoding header at the message level,
  if needed after decoding text parts.

uip/refile.c
/* refile.c -- move or link message(s) from a source folder
 *          -- into one or more destination folders
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include <h/mh.h>
#include <h/utils.h>
#include "../sbr/m_maildir.h"
#include "../sbr/m_mktemp.h"
#include <fcntl.h>

#define REFILE_SWITCHES \
    X("draft", 0, DRAFTSW) \
    X("link", 0, LINKSW) \
    X("nolink", 0, NLINKSW) \
    X("preserve", 0, PRESSW) \
    X("nopreserve", 0, NPRESSW) \
    X("retainsequences", 0, RETAINSEQSSW) \
    X("noretainsequences", 0, NRETAINSEQSSW) \
    X("unlink", 0, UNLINKSW) \
    X("nounlink", 0, NUNLINKSW) \
    X("src +folder", 0, SRCSW) \
    X("file file", 0, FILESW) \
    X("rmmproc program", 0, RPROCSW) \
    X("normmproc", 0, NRPRCSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(REFILE);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(REFILE, switches);
#undef X

static char maildir[BUFSIZ];

struct st_fold {
    char *f_name;
    struct msgs *f_mp;
};

/*
 * static prototypes
 */
static void opnfolds (struct msgs *, struct st_fold *, int);
static void clsfolds (struct st_fold *, int);
static void remove_files (int, char **);
static int m_file (struct msgs *, char *, int, struct st_fold *, int, int, int);
static void copy_seqs (struct msgs *, int, struct msgs *, int);


int
main (int argc, char **argv)
{
    int linkf = 0, preserve = 0, retainseqs = 0, filep = 0;
    int foldp = 0, isdf = 0, unlink_msgs = 0;
    int i, msgnum;
    char *cp, *folder = NULL, buf[BUFSIZ];
    char **argp, **arguments;
    char *filevec[NFOLDERS + 2];
    char **files = &filevec[1];	   /* leave room for remove_files:vec[0] */
    struct st_fold folders[NFOLDERS + 1];
    struct msgs_array msgs = { 0, 0, NULL };
    struct msgs *mp;

    if (nmh_init(argv[0], 1)) { return 1; }

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    /*
     * Parse arguments
     */
    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
	    case AMBIGSW:
		ambigsw (cp, switches);
		done (1);
	    case UNKWNSW:
		adios (NULL, "-%s unknown\n", cp);

	    case HELPSW:
		snprintf (buf, sizeof(buf), "%s [msgs] [switches] +folder ...",
			  invo_name);
		print_help (buf, switches, 1);
		done (0);
	    case VERSIONSW:
		print_version(invo_name);
		done (0);

	    case LINKSW:
		linkf++;
		continue;
	    case NLINKSW:
		linkf = 0;
		continue;

	    case PRESSW:
		preserve++;
		continue;
	    case NPRESSW:
		preserve = 0;
		continue;

	    case RETAINSEQSSW:
		retainseqs = 1;
		continue;
	    case NRETAINSEQSSW:
		retainseqs = 0;
		continue;

	    case UNLINKSW:
		unlink_msgs++;
		continue;
	    case NUNLINKSW:
		unlink_msgs = 0;
		continue;

	    case SRCSW:
		if (folder)
		    adios (NULL, "only one source folder at a time!");
		if (!(cp = *argp++) || *cp == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);
		folder = path (*cp == '+' || *cp == '@' ? cp + 1 : cp,
			       *cp != '@' ? TFOLDER : TSUBCWF);
		continue;
	    case DRAFTSW:
		if (filep > NFOLDERS)
		    adios (NULL, "only %d files allowed!", NFOLDERS);
		isdf = 0;
		files[filep++] = getcpy (m_draft (NULL, NULL, 1, &isdf));
		continue;
	    case FILESW:
		if (filep > NFOLDERS)
		    adios (NULL, "only %d files allowed!", NFOLDERS);
		if (!(cp = *argp++) || *cp == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);
		files[filep++] = path (cp, TFILE);
		continue;

	    case RPROCSW:
		if (!(rmmproc = *argp++) || *rmmproc == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);
		continue;
	    case NRPRCSW:
		rmmproc = NULL;
		continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (foldp > NFOLDERS)
		adios (NULL, "only %d folders allowed!", NFOLDERS);
	    folders[foldp++].f_name =
		pluspath (cp);
	} else
		app_msgarg(&msgs, cp);
    }

    if (!context_find ("path"))
	free (path ("./", TFOLDER));
    if (foldp == 0)
	adios (NULL, "no folder specified");

    /*
     * We are refiling a file to the folders
     */
    if (filep > 0) {
	if (folder || msgs.size)
	    adios (NULL, "use -file or some messages, not both");
	opnfolds (NULL, folders, foldp);
	for (i = 0; i < filep; i++)
	    if (m_file (0, files[i], 0, folders, foldp, preserve, 0))
		done (1);
	/* If -nolink, then "remove" files */
	if (!linkf)
	    remove_files (filep, filevec);
	done (0);
    }

    if (!msgs.size)
	app_msgarg(&msgs, "cur");
    if (!folder)
	folder = getfolder (1);
    strncpy (maildir, m_maildir (folder), sizeof(maildir));

    if (chdir (maildir) == NOTOK)
	adios (maildir, "unable to change directory to");

    /* read source folder and create message structure */
    if (!(mp = folder_read (folder, 1)))
	adios (NULL, "unable to read folder %s", folder);

    /* check for empty folder */
    if (mp->nummsg == 0)
	adios (NULL, "no messages in %s", folder);

    /* parse the message range/sequence/name and set SELECTED */
    for (msgnum = 0; msgnum < msgs.size; msgnum++)
	if (!m_convert (mp, msgs.msgs[msgnum]))
	    done (1);
    seq_setprev (mp);	/* set the previous-sequence */

    /* create folder structures for each destination folder */
    opnfolds (mp, folders, foldp);

    /* Link all the selected messages into destination folders.
     *
     * This causes the add hook to be run for messages that are
     * linked into another folder.  The refile hook is run for
     * messages that are moved to another folder.
     */
    for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++) {
	if (is_selected (mp, msgnum)) {
	    cp = mh_xstrdup(m_name (msgnum));
	    if (m_file (mp, cp, retainseqs ? msgnum : 0, folders, foldp,
                        preserve, !linkf))
		done (1);
	    free (cp);
	}
    }

    /*
     * Update this now, since folder_delmsgs() will save the context
     * but doesn't have access to the folder name.
     */

    context_replace (pfolder, folder);	/* update current folder   */

    /*
     * Adjust "cur" if necessary
     */

    if (mp->hghsel != mp->curmsg
	&& (mp->numsel != mp->nummsg || linkf))
	seq_setcur (mp, mp->hghsel);

    /*
     * Close destination folders now; if we are using private sequences
     * we need to have all of our calls to seq_save() complete before we
     * call context_save().
     */

    clsfolds (folders, foldp);

    /* If -nolink, then "remove" messages from source folder.
     *
     * Note that folder_delmsgs does not call the delete hook
     * because the message has already been handled above.
     */
    if (!linkf) {
	folder_delmsgs (mp, unlink_msgs, 1);
    } else {
	seq_save (mp);	/* synchronize message sequences */
	context_save ();			/* save the context file   */
    }

    folder_free (mp);			/* free folder structure   */
    done (0);
    return 1;
}


/*
 * Read all the destination folders and
 * create folder structures for all of them.
 */

static void
opnfolds (struct msgs *src_folder, struct st_fold *folders, int nfolders)
{
    char nmaildir[BUFSIZ];
    struct st_fold *fp, *ep;
    struct msgs *mp;

    for (fp = folders, ep = folders + nfolders; fp < ep; fp++) {
	if (chdir (m_maildir ("")) < 0) {
	    advise (m_maildir (""), "chdir");
	}
	strncpy (nmaildir, m_maildir (fp->f_name), sizeof(nmaildir));

	/*
	 * Null src_folder indicates that we are refiling a file to
	 * the folders, in which case we don't want to short-circuit
	 * fp->f_mp to any "source folder".
	 */
	if (! src_folder  ||  strcmp (src_folder->foldpath, nmaildir)) {
	    create_folder (nmaildir, 0, done);

	    if (chdir (nmaildir) == NOTOK)
		adios (nmaildir, "unable to change directory to");
	    if (!(mp = folder_read (fp->f_name, 1)))
		adios (NULL, "unable to read folder %s", fp->f_name);
	    mp->curmsg = 0;

	    fp->f_mp = mp;
	} else {
	    /* Source and destination folders are the same. */
	    fp->f_mp = src_folder;
	}

	if (maildir[0] != '\0'  &&  chdir (maildir) < 0) {
	    advise (maildir, "chdir");
	}
    }
}


/*
 * Set the Previous-Sequence and then synchronize the
 * sequence file, for each destination folder.
 */

static void
clsfolds (struct st_fold *folders, int nfolders)
{
    struct st_fold *fp, *ep;
    struct msgs *mp;

    for (fp = folders, ep = folders + nfolders; fp < ep; fp++) {
	mp = fp->f_mp;
	seq_setprev (mp);
	seq_save (mp);
    }
}


/*
 * If you have a "rmmproc" defined, we called that
 * to remove all the specified files.  If "rmmproc"
 * is not defined, then just unlink the files.
 */

static void
remove_files (int filep, char **files)
{
    int i, vecp;
    char **vec, *program;

    /* If rmmproc is defined, we use that */
    if (rmmproc) {
    	vec = argsplit(rmmproc, &program, &vecp);
	files++;		/* Yes, we need to do this */
	for (i = 0; i < filep; i++)
		vec[vecp++] = files[i];
	vec[vecp] = NULL;	/* NULL terminate list */

	fflush (stdout);
	execvp (program, vec);
	adios (rmmproc, "unable to exec");
    }

    /* Else just unlink the files */
    files++;	/* advance past filevec[0] */
    for (i = 0; i < filep; i++) {
	if (m_unlink (files[i]) == NOTOK)
	    admonish (files[i], "unable to unlink");
    }
}


/*
 * Link (or copy) the message into each of the destination folders.
 * If oldmsgnum is not 0, call copy_seqs().
 */

static int
m_file (struct msgs *mp, char *msgfile, int oldmsgnum,
	struct st_fold *folders, int nfolders, int preserve, int refile)
{
    int msgnum;
    struct st_fold *fp, *ep;

    for (fp = folders, ep = folders + nfolders; fp < ep; fp++) {
	/*
	 * With same source and destination folder, don't indicate that
	 * the new message is selected so that 1) folder_delmsgs() doesn't
	 * delete it later and 2) it is not reflected in mp->hghsel, and
	 * therefore won't be assigned to be the current message.
	 */
	if ((msgnum = folder_addmsg (&fp->f_mp, msgfile,
                                     mp != fp->f_mp,
				     0, preserve, nfolders == 1 && refile,
				     maildir)) == -1)
	    return 1;
	if (oldmsgnum) copy_seqs (mp, oldmsgnum, fp->f_mp, msgnum);
    }
    return 0;
}


/*
 * Copy sequence information for a refiled message to its
 * new folder.  Skip the cur sequence.
 */
static void
copy_seqs (struct msgs *oldmp, int oldmsgnum, struct msgs *newmp, int newmsgnum)
{
    char **seq;
    size_t seqnum;

    for (seq = svector_strs (oldmp->msgattrs), seqnum = 0;
	 *seq && seqnum < svector_size (oldmp->msgattrs);
	 ++seq, ++seqnum) {
	if (strcmp (current, *seq)) {
	    assert ((int) seqnum == seq_getnum (oldmp, *seq));
	    if (in_sequence (oldmp, seqnum, oldmsgnum)) {
		seq_addmsg (newmp, *seq, newmsgnum,
                            !is_seq_private (oldmp, seqnum), 0);
	    }
	}
    }
}
